amp/shard_tracker.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/lightningnetwork/lnd/routing/shards"
+)
+
+// Shard is an implementation of the shards.PaymentShards interface specific
+// to AMP payments.
+type Shard struct {
+	child *Child
+	mpp   *record.MPP
+	amp   *record.AMP
+}
+
+// A compile time check to ensure Shard implements the shards.PaymentShard
+// interface.
+var _ shards.PaymentShard = (*Shard)(nil)
+
+// Hash returns the hash used for the HTLC representing this AMP shard.
+func (s *Shard) Hash() lntypes.Hash {
+	return s.child.Hash
+}
+
+// MPP returns any extra MPP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) MPP() *record.MPP {
+	return s.mpp
+}
+
+// AMP returns any extra AMP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) AMP() *record.AMP {
+	return s.amp
+}
+
+// ShardTracker is an implementation of the shards.ShardTracker interface
+// that is able to generate payment shards according to the AMP splitting
+// algorithm. It can be used to generate new hashes to use for HTLCs, and also
+// cancel shares used for failed payment shards.
+type ShardTracker struct {
+	setID       [32]byte
+	paymentAddr [32]byte
+	totalAmt    lnwire.MilliSatoshi
+
+	sharer *SeedSharer
+
+	shards map[uint64]*Child
+}
+
+// A compile time check to ensure ShardTracker implements the
+// shards.ShardTracker interface.
+var _ shards.ShardTracker = (*ShardTracker)(nil)
+
+// NewShardTracker creates a new shard tracker to use for AMP payments. The
+// root shard, setID, payment address and total amount must be correctly set in
+// order for the TLV options to include with each shard to be created
+// correctly.
+func NewShardTracker(root, setID, payAddr [32]byte,
+	totalAmt lnwire.MilliSatoshi) *ShardTracker {
+
+	// Create a new seed sharer from this root.
+	rootShare := Share(root)
+	rootSharer := SeedSharerFromRoot(&rootShare)
+
+	return &ShardTracker{
+		setID:       setID,
+		paymentAddr: payAddr,
+		totalAmt:    totalAmt,
+		sharer:      rootSharer,
+		shards:      make(map[uint64]*Child),
+	}
+}
+
+var zeroShare = Share{}
+
+// NewShard registers a new attempt with the ShardTracker and returns a
+// new shard representing this attempt. This attempt's shard should be canceled
+// if it ends up not being used by the overall payment, i.e. if the attempt
+// fails.
+func (s *ShardTracker) NewShard(pid uint64, last bool) (shards.PaymentShard,
+	error) {
+
+	// If the current share is a zero share, it means we already returned
+	// the last shard.
+	if s.sharer.curr == zeroShare {
+		return nil, fmt.Errorf("splitter exhausted")
+	}
+
+	// Use a random child index.
+	var childIndex [4]byte
+	if _, err := rand.Read(childIndex[:]); err != nil {
+		return nil, err
+	}
+	idx := binary.BigEndian.Uint32(childIndex[:])
+
+	// Depending on whether we are requesting the last shard or not, either
+	// split the current share into two, or return it.
+	var child *Child
+	if last {
```

Comment:
Still getting through the diff, but this seems to assume that the last shard sent will _always_ work? That may not be the case in practice given it's possible that we need to retry that final shard, which means incrementing the child index and trying again in order to not re-use the payment hash.

Reply:
What we'll end up doing us _canceling_ the last shard, which means XOR-ing it together with the current share. Since we here set the current share to zero, that means we are back to the original share.

In the scenario you outline, we'll indeed retry with the same share, but since we are randomizing the 4-byte child index, the payment hash should be different.